src/libs/actions/TeachersUnite.ts

```diff
@@ -0,0 +1,192 @@
+import Onyx, {OnyxEntry, OnyxUpdate} from 'react-native-onyx';
+import * as API from '@libs/API';
+import Navigation from '@libs/Navigation/Navigation';
+import * as OptionsListUtils from '@libs/OptionsListUtils';
+import * as ReportUtils from '@libs/ReportUtils';
+import type {OptimisticCreatedReportAction} from '@libs/ReportUtils';
+import CONST from '@src/CONST';
+import ONYXKEYS from '@src/ONYXKEYS';
+import type {PersonalDetailsList} from '@src/types/onyx';
+
+type CreationData = {
+    reportID: string;
+    reportActionID: string;
+};
+
+type ReportCreationData = Record<string, CreationData>;
+
+type ExpenseReportActionData = Record<string, OptimisticCreatedReportAction>;
+
+let sessionEmail = '';
+let sessionAccountID = 0;
+Onyx.connect({
+    key: ONYXKEYS.SESSION,
+    callback: (value) => {
+        sessionEmail = value?.email ?? '';
+        sessionAccountID = value?.accountID ?? 0;
+    },
+});
+
+let allPersonalDetails: OnyxEntry<PersonalDetailsList>;
+Onyx.connect({
+    key: ONYXKEYS.PERSONAL_DETAILS_LIST,
+    callback: (value) => (allPersonalDetails = value),
+});
+
+function referTeachersUniteVolunteer(partnerUserID: string, firstName: string, lastName: string) {
+    const optimisticPublicRoom = ReportUtils.buildOptimisticChatReport([], CONST.TEACHERS_UNITE.PUBLIC_ROOM_NAME, CONST.REPORT.CHAT_TYPE.POLICY_ROOM, CONST.TEACHERS_UNITE.POLICY_ID);
+    const optimisticData: OnyxUpdate[] = [
+        {
+            onyxMethod: Onyx.METHOD.SET,
+            key: `${ONYXKEYS.COLLECTION.REPORT}${optimisticPublicRoom.reportID}`,
+            value: {
+                ...optimisticPublicRoom,
+                reportID: optimisticPublicRoom.reportID,
+                policyName: CONST.TEACHERS_UNITE.POLICY_NAME,
+            },
+        },
+    ];
+
+    type ReferTeachersUniteVolunteerParams = {
+        publicRoomReportID: string;
+        firstName: string;
+        lastName: string;
+        partnerUserID: string;
+    };
+
+    const parameters: ReferTeachersUniteVolunteerParams = {
+        publicRoomReportID: optimisticPublicRoom.reportID,
+        firstName,
+        lastName,
+        partnerUserID,
+    };
+
+    API.write('ReferTeachersUniteVolunteer', parameters, {optimisticData});
+    Navigation.dismissModal(CONST.TEACHERS_UNITE.PUBLIC_ROOM_ID);
+}
+
+/**
+ * Optimistically creates a policyExpenseChat for the school principal and passes data to AddSchoolPrincipal
+ */
+function addSchoolPrincipal(firstName: string, partnerUserID: string, lastName: string) {
+    const policyName = CONST.TEACHERS_UNITE.POLICY_NAME;
+    const policyID = CONST.TEACHERS_UNITE.POLICY_ID;
+    const loggedInEmail = OptionsListUtils.addSMSDomainIfPhoneNumber(sessionEmail);
+    const reportCreationData: ReportCreationData = {};
+
+    const expenseChatData = ReportUtils.buildOptimisticChatReport([sessionAccountID], '', CONST.REPORT.CHAT_TYPE.POLICY_EXPENSE_CHAT, policyID, sessionAccountID, true, policyName);
```

Comment:
maybe you can add type to this const?

Reply:
@kubabutkiewicz it's typed automatically, do you think it makes sense to type it additionally?
![image](https://github.com/Expensify/App/assets/23176449/c03f578a-2210-4b2d-9330-8e4038ff7f67)